rfc/0003-indexmanagement.md

```diff
@@ -0,0 +1,156 @@
+ - `createPrimaryIndex`: create a primary index on the current Bucket.
+ - `createIndex`: create a secondary GSI index on the current Bucket.
+ - `dropPrimaryIndex`: drop the current Bucket's primary index.
+ - `dropIndex`: drop a specific secondary index of the current Bucket.
+ - `buildDeferredIndexes`: trigger the build of indexes that were created in a deferred fashion (see below).
+
+Additionally it is open for discussion wether or not this RFC should also include the following ***2*** methods in a cross-SDK fashion, or if they should be optional:
+ - `watchIndex`: poll the `system:indexes` N1QL view until a given index changes its status to "online" (or a maximum amount of time has passed).
+ - `watchIndexes`: poll the `system:indexes` N1QL view until a given list of index all become "online" (or a maximum amount of time has passed).
+
+
+The underlying design would be to rely on existing `N1QL` verbs and, for the listing and polling part, to use the `system:indexes` keyspace. Each method will internally create and execute a specific N1QL query according to its parameter and the name of the current Bucket.
+
+Each method will offer a bit of tuning with a couple of parameters, so let's detail a more complete signature for each one.
+
+## Listing Indexes
+Signature:
+```java
+List<IndexInfo> listIndexes()
+```
+
+Remarks:
+The method relies on `system:indexes` with the following N1QL statement:
+```sql
+SELECT idx.* FROM system:indexes AS idx WHERE keyspace_id = `BUCKET_NAME_INJECTED_HERE` ORDER BY is_primary DESC, name ASC;
+```
+
+For SDK in language with no dynamic typing, each row will be converted to a `IndexInfo` object with the following attributes:
+
+```java
+String name
+boolean isPrimary
+IndexType type //enum type GSI/VIEW
+String rawType //raw string type in case a new index type is introduced, for forward-compatibility
+String state //eg. "online", "pending"
+String keyspace
+String namespace
+JsonArray indexKey //or equivalent Json representation
+```
+
+## Creating Indexes
+There are two specificities for index creation (both of primary and secondary indexes): the index could already exist and there is the possibility of deferring the build.
+
+In the first specificity, it makes sense to offer two alternatives: either this is considered an error (as N1QL does) or the creation should be purely and simply ignored.
+This choice can be made through a first `boolean` parameter, `ignoreIfExist`. If it is `false`, then a SDK-related exception should be raised (eg. in Java a `CouchbaseException`).
+
+The second specificity can be triggered by another `boolean` parameter, `defer`. If set to true, the N1QL query will use the "with defer" syntax and the index will simply be "pending".
+
+Signatures:
+```java
+boolean createPrimaryIndex(boolean ignoreIfExist, boolean defer)
+
+boolean createIndex(String indexName, List<String> fields, boolean ignoreIfExist, boolean defer)
+```
```

Comment:
I think we need an overload here for complex indexes. For example, index hints:

    CREATE INDEX `pv_ndx_02` ON `default`(`userId`,(meta().`id`),`type`) 
    WHERE ((`type` = 'pv') 
    AND ((to_number(`userId`) % 2) = 0)) USING GSI;

Unless of course, if by "fields" you meant "indexes"...